### PR TITLE
Support for Docker Registry 2.0 notifications (by HolidayCheck AG.) (JENKINS-30931)

### PR DESCRIPTION
**Pull request by lguminski**
The PR adds compatibility with Docker Registry 2.0. It keeps also full backwards compatibility with Docker Hub notifications.

After merging the plugin will accept notifications from Docker Hub and Docker Registry
* Docker Hub at `http://JENKINS/dockerhub-webhook/notify` 
* Docker Registry 2.0 at `http://JENKINS/dockerregistry-webhook/notify`

One en event gets accepted, the rest of logic remains the same. That's why I think that extending this plugin is a better idea than creating a new one.

### CODE REVIEW

**Comment by GrantStreetGroup** [comment]
I'm looking forward to using this feature...  please have it merged.

**Comment by rwincewicz** [comment]
Now that the Docker Registry supports notifications this would be a great addition to the plugin. This would fit perfectly into our docker build + deploy workflow.

**Comment by GrantStreetGroup** [comment]
Any idea how to 'build' this plugin for jenkins and use it?   save waiting for JenkinsCI to do something with it?

**Comment by michaelneale** [comment]
@GrantStreetGroup cloning it and running "mvn hpi:run" should fire up a jenkins with the plugin running and you can try it out.

File: src/main/java/org/jenkinsci/plugins/registry/notification/webhook/JSONWebHook.java [hunk 1]
```diff
@@ -21,55 +21,33 @@
...
-import hudson.Extension;
-import hudson.Main;
-import hudson.model.Action;
-import hudson.model.CauseAction;
-import hudson.model.Job;
-import hudson.model.ParameterDefinition;
-import hudson.model.ParameterValue;
-import hudson.model.ParametersAction;
-import hudson.model.ParametersDefinitionProperty;
+import hudson.model.*;
```
Comment (michaelneale) [inline src/main/java/org/jenkinsci/plugins/registry/notification/webhook/JSONWebHook.java]: Should this be specific classes not a wildcard?

**Comment by michaelneale** [comment]
Wow this is a good big one. For those not following along - dockerhub (the public one docker inc run) and docker registry (the one people run privately) have a different notification api. The dockerhub has a simple webhook on push events, where as the registry has a fine grained one. The plugin as it was didn't support the private registry. This is kind of a rewrite so it can take notifications from registry, and commit webhooks from dockerhub into the same place (at least that is how I read it). I really like it, thanks @lguminski. Would like others to look at it as it is fairly large, but makes sense hopefully.

**Comment by michaelneale** [comment]
@lguminski are you using this with your private registry or dockerhub right now?

**Comment by michaelneale** [comment]
Also, this PR implies a plugin name change (at least in display names) to be "docker registry" vs "Docker Hub" in my opinion (but not necessary for it to be accepted).

**Comment by michaelneale** [comment]
LGTM - waiting for some others to review.

**Comment by michaelneale** [comment]
@lguminski I wasn't able to verify the private registry trigger worked. The command I used was: 

```
curl -X POST -H "Content-Type: application/json" http://localhost:8080/jenkins/dockerregistry-webhook/notify -d @src/test/resources/private-registry-payload-1-repository.json
```

and the job was set to trigger on: 

```
jplock
zookeeper
busybox
library
library/busybox
jplock/zookeeper
```

This triggers on busybox changes using the public dockerhub json and endpoint, but not private. Tests pass for me but I can't seem to verify it working on the "real" data.

**Comment by GrantStreetGroup** [comment]
@michaelneale - Thanks,  as I'm build a docker container for Jenkins, I can add 'maven' to it for building the plugin.   

It's out intention to use the docker registry 2.0 notifications to trigger a jenkins deployment to test. 

Will let you know how we get on.

-- Steven

**Comment by GrantStreetGroup** [comment]
@lguminski   What do you put into the Jenkins job settings for this?
-- Steven

**Comment by michaelneale** [comment]
@GrantStreetGroup you can also mvn install - to generate a .hpi plugin file, and include that in your image in the plugins directory (probably better than running in production via hpi:run, its only for development time).

**Comment by GrantStreetGroup** [comment]
@michaelneale   Got the package installed OK...  Just need to understand how to setup the jenkins side in a job..

**Comment by michaelneale** [comment]
@GrantStreetGroup the /jenkins is only needed when running via mvn hpi:run. When installed in a production jenkins, yes, you don't want that path (just a quirk of plugin development).

**Comment by GrantStreetGroup** [comment]
It does not seem to be working...    Just trying to get a handle on how the payload maps to the item configuration.

Steven

**Comment by GrantStreetGroup** [comment]
OK after much reading of the code:  

'''
dummy-job - #3 Triggered by push of docker-registry.grantstreet.com:5000/grantst/busybox to DockerRegistry@docker-registry.grantstreet.com:5000 (Open)
'''

(via notification to slack)

However this was a 'pull' notification not a push notification.   The trigger really needs to not happen on a pull.

**Comment by GrantStreetGroup** [comment]
OK,  here is my considered opinion:

DockerHub notifications and Docker Registry notifications are very different systems, this implementation lacks a number of features to be truly ready of prime time. (IMHO)

Currently any 'event' for a matching repository will trigger a build.

* Needs to differentiate between push and pull notifications, so only push notifications trigger
* Needs to trigger on the push of the manifest, not just individual layers.

One feature that would be great, is being able to tell which 'tag' had been pushed, so the trigger can be set to tag level of an image.

**Comment by michaelneale** [comment]
Yes, it looks like there still are some issues with this PR

**Pull request merged**